go/libkb/pvl_interp.go

```diff
@@ -0,0 +1,825 @@
+	HTML   *goquery.Document
+	JSON   *jsonw.Wrapper
+}
+
+type PvlMode string
+
+const (
+	PvlModeJSON   PvlMode = "json"
+	PvlModeHTML   PvlMode = "html"
+	PvlModeString PvlMode = "string"
+	PvlModeDNS    PvlMode = "dns"
+)
+
+const (
+	PvlAssertRegexMatch    = "assert_regex_match"
+	PvlAssertFindBase64    = "assert_find_base64"
+	PvlWhitespaceNormalize = "whitespace_normalize"
+	PvlRegexCapture        = "regex_capture"
+	PvlFetch               = "fetch"
+	PvlSelectorJSON        = "selector_json"
+	PvlSelectorCSS         = "selector_css"
+	PvlTransformURL        = "transform_url"
+)
+
+type PvlStep func(GlobalContextLite, *jsonw.Wrapper, PvlScriptState) (PvlScriptState, ProofError)
+
+var PvlSteps = map[string]PvlStep{
+	PvlAssertRegexMatch:    pvlStepAssertRegexMatch,
+	PvlAssertFindBase64:    pvlStepAssertFindBase64,
+	PvlWhitespaceNormalize: pvlStepWhitespaceNormalize,
+	PvlRegexCapture:        pvlStepRegexCapture,
+	PvlFetch:               pvlStepFetch,
+	PvlSelectorJSON:        pvlStepSelectorJSON,
+	PvlSelectorCSS:         pvlStepSelectorCSS,
+	PvlTransformURL:        pvlStepTransformURL,
+}
+
+// Checkproof verifies one proof by running the pvl on the provided proof information.
+func CheckProof(g GlobalContextLite, pvl *jsonw.Wrapper, service keybase1.ProofType, link RemoteProofChainLink, h SigHint) ProofError {
+	if perr := pvlValidateChunk(pvl, service); perr != nil {
+		return perr
+	}
+
+	sigBody, sigID, err := OpenSig(link.GetArmoredSig())
+	if err != nil {
+		return NewProofError(keybase1.ProofStatus_BAD_SIGNATURE,
+			"Bad signature: %v", err)
+	}
+
+	scripts, perr := pvlChunkGetScripts(pvl, service)
+	if perr != nil {
+		return perr
+	}
+
+	newstate := func() PvlScriptState {
+		vars := PvlScriptVariables{
+			UsernameService: link.GetRemoteUsername(), // Blank for DNS-proofs
+			UsernameKeybase: link.GetUsername(),
+			Sig:             sigBody,
+			SigIDMedium:     sigID.ToMediumID(),
+			SigIDShort:      sigID.ToShortID(),
+			Hostname:        link.GetHostname(), // Blank for non-{DNS/Web} proofs
+		}
+
+		// Enforce prooftype-dependent variables.
+		webish := (service == keybase1.ProofType_DNS || service == keybase1.ProofType_GENERIC_WEB_SITE)
+		if webish {
+			vars.UsernameService = ""
+		} else {
+			vars.Hostname = ""
+		}
+
+		state := PvlScriptState{
+			PC:           0,
+			Service:      service,
+			Vars:         vars,
+			ActiveString: h.apiURL,
+			FetchURL:     h.apiURL,
+			HasFetched:   false,
+			FetchResult:  nil,
+		}
+		return state
+	}
+
+	if service == keybase1.ProofType_DNS {
+		perr = pvlRunDNS(g, scripts, newstate())
+		if perr != nil {
+			return perr
+		}
+		return nil
+	}
+
+	// Run the scripts in order.
+	// If any succeed, the proof succeeds.
+	// If one fails, the next takes over.
+	// If all fail, report only the first error.
+	var errs []ProofError
+	for _, script := range scripts {
+		perr = pvlRunScript(g, script, newstate())
+		if perr == nil {
+			return nil
+		}
+		errs = append(errs, perr)
```

Comment:
Logging here is probably important, since we're not going to be able to return all of these. Alternatively, we could define a compound error type that concatenates the error messages from all the failures together. (Reporting just the first error might be confusing in a situation where a proof was intended to match the second script, and failed because of a spelling error there or something. There the second error would be much more relevant, but we have no way of knowing that.)